### ISSUE TITLE
stats file renaming crashes sporadically on Windows

### ISSUE THREAD

**Issue by mumpitzstuff** [ccache/ccache#537]
### How to reproduce ###

1. Can not be reproduced step by step.

### Actual behavior ###

I see the following error and the build fails sporadically (always the stats temp file):

ccache: error: failed to rename C:/workspace/PRP_RH850/.ccache/1/stats.tmp.unknown.736.SS92c3 to C:/workspace/PRP_RH850/.ccache/1/stats

### Expected behavior ###

Any problems with the stats file should not result in a crash of the build. But the root cause should also be eliminated.

### Environment ###

ccache version 3.7.7 from here:
https://github.com/cristianadam/ccache/actions/runs/35652485

Compiler is started in parallel by our make (4 threads = 4 cores).

Windows

**Comment by mumpitzstuff** [comment]
Example 1 (history of single problem):

[.860907 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.860907 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1000 microseconds
[.862907 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.862907 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.862907 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 2000 microseconds
[.865904 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.865904 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.865904 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 4000 microseconds
[.870901 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.870901 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.870901 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 8000 microseconds
[.879896 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.879896 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.879896 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 16000 microseconds
[.896889 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.896889 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.896889 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 32000 microseconds
[.929868 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.929868 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 64000 microseconds
[.994832 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.994832 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 128000 microseconds
[.123758 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.123758 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.123758 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 256000 microseconds
[.380611 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.380611 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.380611 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 512000 microseconds
[.892963 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.040565 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.041564 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1024000 microseconds
[.065808 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.065808 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.065808 8480 ] lockfile_acquire: breaking D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock


Example 2 (we get more and more of these problems!!!):

[.436580 42772] Looking for result name in D:\Jenkins\workspace\L1\test\.ccache/2/4/37a0b4d27714e2106395bb5fd14d8488796d11.manifest
[.471600 6608 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.471600 6608 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.471600 6608 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 128000 microseconds
[.484125 33864] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.484125 33864] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.484125 33864] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 512000 microseconds
[.587116 20820] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.587116 20820] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.587116 20820] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 512000 microseconds
[.600114 6608 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.892963 8480 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.040565 6608 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.040565 33864] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.040565 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
econds
[.041564 33864] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1024000 microseconds
[.041564 8480 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1024000 microseconds
[.100134 20820] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.100134 20820] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.100134 20820] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1024000 microseconds
[.298212 6608 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.298212 6608 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.298212 6608 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 512000 microseconds
[.811129 6608 ] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.811129 6608 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.811129 6608 ] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock; sleeping 1024000 microseconds



Example 3 (this is the final crash):

[.065808 33864] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: File exists
[.065808 33864] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
[.065808 33864] lockfile_acquire: breaking D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
[.065808 33864] lockfile_acquire: open WRONLY D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock: File exists
[.066808 33864] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock: 
[.066808 33864] lockfile_acquire: failed to acquire D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock; sleeping 1000 microseconds
[.069806 33864] Acquired lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock
[.069806 33864] Releasing lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
[.069806 33864] Unlink D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
[.069806 33864] Releasing lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock
[.069806 33864] Unlink D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock.lock
[.071805 33864] Acquired lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
[.073804 33864] can't rename file D:\Jenkins\workspace\L1\test\.ccache/f/stats.tmp.unknown.33864.iG2Kb7 to D:\Jenkins\workspace\L1\test\.ccache/f/stats OS returned error: ../src/legacy_util.cpp failed with error 183: Cannot create a file when that file already exists.

**Comment by mumpitzstuff** [comment]
In my opinion it is not a good idea to just double the timeout on each fail and to use such a high timeout value. Our project has ~3300 files and the build is finished in 2 minutes if all files already exists in the cache. This means round about 28 files per second or 28 times per second acquiring a lock for the stats file. The chance is really really high that multiple processes try to acquire a lock at the same time. **Therefore i would use a much lower random timeout value (e.g. random value between 50 and 250 milliseconds could be sufficient).** But in the end this is just a deadlock avoidance and no prevention...

**Comment by mumpitzstuff** [comment]
I modified lockfile.c a litte bit to give it a try:

```
unsigned random_sleep = 0; // Microseconds.
.
.
.
srand(time(NULL));
random_sleep = to_sleep + (rand() % 1024); // multiply of 8 can be optimized by the compiler
cc_log("lockfile_acquire: failed to acquire %s; sleeping %u microseconds",
		lockfile, random_sleep);
usleep(random_sleep);
slept += random_sleep;
```
Now i use random numbers as timeout values 1000 - 2024 and the timeout value is not doubled on each fail.

**Comment by jrosdahl** [comment]
I think that there are at least four things to discuss here:

#### 1. Lock contention when many processes race to update the 16 stats files

> Our project has ~3300 files and the build is finished in 2 minutes if all files already exists in the cache. This means round about 28 files per second or 28 times per second acquiring a lock for the stats file.

You have to divide by 16 as well since there are 16 stats files, so it's around 2 per second per stats file. That's not very frequent; I've seen 10 times as much without noticable problems.

That said, I plan to improve this in ccache 4.0 by having more stats files and thus reducing the probability of lock contention.

#### 2. Some problem with locking on Windows

Have a look at the "lock info" rows in example 1. They all say "unknown:48776:1580918461" which means that process 8480 is waiting for process 48776 to release the lock, but it never happens and 8480 finally breaks the lock after 2 seconds. Having another back-off strategy wouldn't help in this case since the process is just waiting for 2 seconds to pass. And if a process fails to update the stats within 2 seconds, something else is fishy. Either process 48776 died before releasing the lock or it failed to release the lock (that is, just delete the file). It's hard to say since there is no full log file to read.

I'm not against improving the back-off algorithm since it's indeed a bit naive, but I don't think that it would fix the problem you experience.

In example 2 the lock info is identical to those in example 1. "1580918461" is the timestamp when the lockfile was created and since it's the same in example 1 and 2 I assume both log snippets are from the same run, only filtered differently?

#### 3. Crash on Windows when trying to rename a temporary file to the final stats file

> can't rename file D:\Jenkins\workspace\L1\test\.ccache/f/stats.tmp.unknown.33864.iG2Kb7 to D:\Jenkins\workspace\L1\test\.ccache/f/stats OS returned error: ../src/legacy_util.cpp failed with error 183: Cannot create a file when that file already exists.

I guess this means that the logic around locking or lock breaking has failed since the destination file should haven been deleted before renaming it into place. (This error can't happen on non-Windows platforms.) A failure should of course not lead to a hard crash, so here's room for improvement.


#### 4. "ccache 3.7.7"

> ../src/legacy_util.cpp

Wait a minute, I thought you said you run ccache 3.7.7 but ccache 3.7.7 doesn't have a src/legacy_util.cpp file.

> ccache version 3.7.7 from here:
https://github.com/cristianadam/ccache/actions/runs/35652485

Oh. I see. cristianadam/ccache seems to have a v3.7.7 tag based on _master_ plus additional experimental Visual C++ support! That's not good.

@cristianadam: I hope that this is just a misunderstanding. The 3.7.x releases are cut from the 3.7-maint branch, not master. More info here: https://ccache.dev/repo.html. master currently has 10 months of pre-4.0 development that is not ready for release, so calling an in-heavy-development revision "3.7.7" is misleading at best. Please consider removing the v3.7.6 and v3.7.7 tags since they seem to be based on pre-4.0 master, not the real 3.7.6 and 3.7.7 versions. Regarding Visual C++ support: I would prefer not to receive bug reports for code that is not yet part of any ccache branch. Thanks for your consideration.

**Comment by cristianadam** [comment]
Bummer, I was wondering if the c++fication was still part of 3.7.x series, I was wrong.

Since the releases I've made are used in 'production' I'll add a bit fat warning on the release page. I just can't remove them.

For the Visual C++bug reports, you can point them to my fork.

**Comment by mumpitzstuff** [comment]
We used the official 3.7.7 release before but we had some trouble with random crashes. Therefore we used a newer version but also got random crashes all the time. Now we tried to get the root cause and found the problem with the stats file locking. This part of the code is nearly the same in both versions i think, therefore it should not make any differences. Now we switched back to the official 3.7.7 version + the small patches seen above. We also disabled the stats generation, just to be sure. :)
The problem only occur under heavy load (many direct cache hits) with multicore compilation (up to 8 ccache/compiler calls in parallel!!!). Round about 1 of 15-20 builds will crash.

Sry that i can not provide the whole logfile but it contains some stuff which i do not want to make public. But i could send it per mail if needed.

**Comment by jrosdahl** [comment]
@cristianadam wrote:
> Since the releases I've made are used in 'production' I'll add a bit fat warning on the release page. I just can't remove them.

That's unfortunate.

> For the Visual C++bug reports, you can point them to my fork.

It looks like your fork has no bug tracker so how should users report bugs?

If you have decided to fork your own version of ccache and do releases with new functionality (not only builds of ccache versions, which I previously thought that your fork was about), please make that clear by using some other name than ccache for the project, README, binaries and documentation.

Or instead of forking, help me and the ccache project forward with Windows stuff as I asked for in #447. Sounds reasonable?

**Comment by jrosdahl** [comment]
@mumpitzstuff wrote:
> [...]  We also disabled the stats generation, just to be sure. :)

Yes, that will avoid the problem since the lockfile routines are only used for stats files. :slightly_smiling_face:

To solve the real problem somebody needs to stress test the lockfile routines on Windows and fix any issues. Help would be appreciated!

**Comment by cristianadam** [comment]
> 
> > For the Visual C++bug reports, you can point them to my fork.
> 

I've enabled Issues on at my fork now. 

> It looks like your fork has no bug tracker so how should users report bugs?
> 
> If you have decided to fork your own version of ccache and do releases with new functionality (not only builds of ccache versions, which I previously thought that your fork was about), please make that clear by using some other name than ccache for the project, README, binaries and documentation.
> 
> Or instead of forking, help me and the ccache project forward with Windows stuff as I asked for in #447. Sounds reasonable?

I don't want to fork fork ccache. I just need to have binaries and Visual C++ support.  When ccache will have these I won't have a reason for keeping my fork.

**Comment by mumpitzstuff** [comment]
We did not got a crash again with the original 3.7.7 ccache version and disabled stats file generation. Therefore i am pretty sure that the reason for our problems was the stats file locking.
May be this simple patch in lockfile.c could fix the problem:
```
#ifndef _WIN32
		to_sleep *= 2;
#endif
```
This means ccache will try to acquire the lock much more often than just 11 times in case of stats file locking. To disable stats generation as default and to insert a hint that stats generation in production environments can lead to problems because of excessive locking could help also to avoid this problem.

**Comment by jrosdahl** [comment]
@mumpitzstuff wrote:
> We got no crashes the last 3 days with the original 3.7.7 ccache version and disabled stats file generation.

Good to hear that disabling stats worked around the problem as expected.

> May be this simple patch in lockfile.c could fix the problem:
> 
> ```
> #ifndef _WIN32
> 		to_sleep *= 2;
> #endif
> ```
> 
> This means ccache will try to acquire the lock much more often than just 11 times in case of stats file locking.

Thanks for the suggestion. However, that will neither fix the problem nor make it less frequent. Or at least not the problem indicated by the log that you provided – trying to acquire the lock more often won't help when the process holding the lock doesn't release it (as described in part 2 of my [previous comment](https://github.com/ccache/ccache/issues/537#issuecomment-586577551)).

> To disable stats generation as default and to insert a hint that stats generation in production environments can lead to problems because of excessive locking could help also to avoid this problem.

I don't think that disabling stats generation by default is a good option. And from what I have seen the problem does not seem to be due to excessive locking (as in: lock contention causing timeouts); two updates per second on average is not much. What needs to be done is to investigate and fix the assumed locking issue mentioned in [part 3](https://github.com/ccache/ccache/issues/537#issuecomment-586577551).

That said, I suspect that the crash part of the problem could have been fixed on master by the use of `MoveFileExA(..., MOVEFILE_REPLACE_EXISTING);)` in 743f35a1299bf6b412fd78e04ff3cadf2e18a04f. Backported to 3.7-maint in 14ef13c2325e now. It doesn't fix the locking algorithm glitch, though.

> Sry that i can not provide the whole logfile but it contains some stuff which i do not want to make public. But i could send it per mail if needed.

You can send the log to me (joel@rosdahl.net) if you like. In particular, it would be interesting to know what process 48776 logged since that's the process that was supposed to release the lock but didn't.

**Comment by mumpitzstuff** [comment]
I have send some logfiles to you and will try to use your patch the next days.

**Comment by jrosdahl** [comment]
> This seems to be the relevant part you are looking for:

Yes, exactly. That's what started the chain of events.

Expanded with more surrounding logs and some commentary:

```
[.848914 48776] Result: cache hit (direct)
[.852912 48776] Acquired lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
```

So 48776 wants to update a counter in f/stats. Meanwhile:

```
[.855910 8480 ] Unlink D:\Jenkins\workspace\L1\test\.ccache_tmp/tmp.stderr.unknown.8480.P1kAlM
[.855910 8480 ] Succeeded getting cached result
[.855910 8480 ] Result: cache hit (direct)
```

8480 also wants to update a counter in f/stats.

```
[.858908 48776] Releasing lock D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
```

48776 is now done. But:

```
[.858908 48776] Unlink D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock
\.ccache/f/stats.lock: File exists
[.858908 48776] Unlink failed: Permission denied
```

Why did 48776 fail to delete the file? 8480 presumably tried to open the file at the same time but failed like expected since the file still exists:

```
[.860907 8480 ] lockfile_acquire: lock info for D:\Jenkins\workspace\L1\test\.ccache/f/stats.lock: unknown:48776:1580918461
```

So I think this is what happened:

* 8480's `open(O_CREAT | O_EXCL)` failed.
* 8480 then opened the file for reading its content (to check if it's stale).
* While 8480 had the file open for reading, 48776's unlink call failed.

I think that the fix (or workaround if you will) here is to retry the `unlink` until it succeeds.

Regardless, the lock breaking algorithm must work properly as well in case a stale lock file is left for any reason, such as a process being terminated while holding the lock. If two processes have decided to break the lock at the same time then we have to accept that one of the counter updates is lost, which I think is fulfilled on Windows as well now with 14ef13c.

**Comment by mumpitzstuff** [comment]
This patch is running since last friday now and stats generation is activated again. Please give me some days to verify that everything is running fine now.
Thank you for your analysis!

**Comment by mumpitzstuff** [comment]
Got another crash today (patch was used). This could be the root cause but i am not sure:

```
[2020-02-25T13:31:05.628691 50352] Succeeded getting cached result
[2020-02-25T13:31:05.628691 50352] Result: cache hit (direct)
[2020-02-25T13:31:05.631690 50352] lockfile_acquire: open WRONLY <path>/.ccache/e/stats.lock: File exists
[2020-02-25T13:31:05.633689 50352] lockfile_acquire: lock info for <path>/.ccache/e/stats.lock: FE-Z1L6C:5440:1582633865
[2020-02-25T13:31:05.633689 50352] lockfile_acquire: failed to acquire <path>/.ccache/e/stats.lock; sleeping 1000 microseconds
[2020-02-25T13:31:05.634688 5440 ] Releasing lock <path>/.ccache/e/stats.lock
[2020-02-25T13:31:05.634688 5440 ] Unlink <path>/.ccache/e/stats.lock
[2020-02-25T13:31:05.635687 50352] lockfile_acquire: open WRONLY <path>/.ccache/e/stats.lock: File exists
[2020-02-25T13:31:05.636687 50352] lockfile_acquire: open RDONLY <path>/.ccache/e/stats.lock: Permission denied
[2020-02-25T13:31:05.636687 50352] Failed to acquire lock <path>/.ccache/e/stats.lock
```

**Comment by mumpitzstuff** [comment]
[https://github.com/jrosdahl/ccache/tree/improve-windows-lockfiles](url)

We have not experienced a single crash since we started using this software (with stats enabled). The problem seems to be fixed.

**Comment by jrosdahl** [comment]
@mumpitzstuff: Thanks for testing!
